src/services/blocks/trace/Trace.ts

```diff
@@ -0,0 +1,690 @@
+			// Since we assume the `spanIds` are for spans from an apply extrinsic action group,
+			// we assume there is always an `:extrinsic_index` event
+			throw new InternalServerError('Expected at least one extrinsic index');
+		}
+
+		if (extrinsicIndex.length > 1) {
+			// If there are multiple reads we check that they are all the same. If they
+			// are not this program has some incorrect assumptions and this error should
+			// be reported to the maintainers.
+			for (const [i, extIdx] of extrinsicIndex.entries()) {
+				if (i > 0 && !extrinsicIndex[i - 1]?.eq(extIdx)) {
+					throw new InternalServerError(
+						'Expect extrinsic to only be applied at a single index.'
+					);
+				}
+			}
+		}
+
+		return extrinsicIndex[0];
+	}
+
+	/**
+	 * Find the Ids of all the spans which are descendant of the span `root`.
+	 *
+	 * @param root span which we want all the descendants of
+	 * @param spansById map of span id => `SpanWithChildren`
+	 */
+	private static findDescendants(root: number, spansById: SpansById): number[] {
+		// Note: traversal order doesn't matter here
+		const stack = spansById.get(root)?.children;
+		if (!stack) {
+			return [];
+		}
+
+		const descendants = [];
+		while (stack.length) {
+			const curId = stack.pop();
+			if (!curId) {
+				break;
+			}
+
+			descendants.push(curId);
+
+			const curSpanChildren = spansById.get(curId)?.children;
+			curSpanChildren && stack.push(...curSpanChildren);
+		}
+
+		return descendants;
+	}
+
+	/**
+	 * Parses spans by
+	 * 1) creating a Map of span's `id` to the the span with its children
+	 * 2) creating an array with the same spans
+	 *
+	 * @param spans spans to parse
+	 */
+	private static parseSpans(
+		spans: TraceSpan[]
+	): {
+		spansById: SpansById;
+		parsedSpans: SpanWithChildren[];
+		executeBlockSpanId: number;
+	} {
+		const spansById = new Map<number, SpanWithChildren>();
+		const parsedSpans = [];
+		let executeBlockSpanId;
+		// This loop does 3 things: 1) build `spansById` 2) build `parseSpans` 3) finds
+		// `executeBlockSpanId`
+		for (const span of spans) {
+			const spanWithChildren = { ...span, children: [] };
+
+			if (
+				span.name === SPANS.executeBlock.name &&
+				span.target === SPANS.executeBlock.target
+			) {
+				if (executeBlockSpanId !== undefined) {
+					throw new InternalServerError(
+						'More than 1 execute block span found when only 1 was expected'
+					);
+				}
+				executeBlockSpanId = span.id;
+			}
+
+			parsedSpans.push(spanWithChildren);
+			spansById.set(span.id, spanWithChildren);
+		}
+
+		if (executeBlockSpanId === undefined) {
+			throw new InternalServerError('execute_block span could not be found');
+		}
+
+		// Go through each span, and add its `id` to the `children` array of its
+		// parent span, creating a tree of spans starting with `executeBlock` span as
+		// the root.
+		for (const span of spans) {
+			if (!span.parentId) {
+				continue;
+			}
+
+			if (!spansById.has(span.parentId)) {
```

Comment:
same question here if `has() + get()` could be replace with `get()` and check !== undefined?